### ISSUE TITLE
ganesha 2.6.3 Win7 client new or copy folder error

### ISSUE THREAD

**Issue by uyghfjhh** [nfs-ganesha/nfs-ganesha#381]
With win7 client and ganesha 2.6.3 FSAL_CEPH, Error when creating folders and copying folders

I sniffed the network traffic and found that when using the win7 client, in the mkdir operation, atime needs to use the client time, as shown below:

![win-mkdir-req](https://user-images.githubusercontent.com/13869600/49924620-d0e45b00-fef1-11e8-8e77-7b23823acd4d.PNG)

When ganesha processes this mkdir request, it does not return any attributes to the server (as shown below), which eventually causes the client to report an error.

![win-mkdir-res](https://user-images.githubusercontent.com/13869600/49924593-b90cd700-fef1-11e8-8b35-6b01a0264b2b.PNG)

But there is no such problem when using FSAL_VFS。

I compared the code of FSAL_VFS, then modified the code of FSAL_CEPH, and then I can create and copy the folder without error

```c
diff --git a/src/FSAL/FSAL_CEPH/handle.c b/src/FSAL/FSAL_CEPH/handle.c
old mode 100644
new mode 100755
index f7e9a27..27b3bd3
--- a/src/FSAL/FSAL_CEPH/handle.c
+++ b/src/FSAL/FSAL_CEPH/handle.c
@@ -282,7 +282,13 @@ static fsal_status_t ceph_fsal_mkdir(struct fsal_obj_handle
                                     fsal_err_txt(status));
                        (*new_obj)->obj_ops->release(*new_obj);
                        *new_obj = NULL;
-               }
+               } else if (attrs_out != NULL) {
+                status = (*new_obj)->obj_ops->getattrs(*new_obj, attrs_out);
+                if (FSAL_IS_ERROR(status) && (attrs_out->request_mask & ATTR_RD
+                     LogFullDebug(COMPONENT_FSAL, "getattrs status=%s", fsal_er
+                }
+        }
+
        } else {
                status = fsalstat(ERR_FSAL_NO_ERROR, 0);
```

Is this a bug in ganesha?

**Comment by dang** [comment]
This needs to be looked at, I think.  I checked CEPH, GLUSTER, and VFS, 
and they all three do different things for attrs_out.  CEPH ignores it 
(definitely a bug); GLUSTER sets it from the stat results (ignoring 
anything set in attrib) and VFS does a full getattrs().

Daniel

**Comment by jtlayton** [comment]
I think the bug is in ceph_fsal_mkdir. In the attrib->valid_mask case, we don't populate attrs_out. Probably we should just do a ceph_ll_getattr to get the attributes after the setattr2 call. In almost all cases, those attrs should be in libcephfs' cache...

**Comment by jtlayton** [comment]
I think this patch will likely fix it, but it's not yet tested. @uyghfjhh : If you have the ability to help test it, that would be helpful:

https://github.com/jtlayton/nfs-ganesha/commit/d83f9dace2314eb9697998c9741b224956ecdfd6

**Comment by jtlayton** [comment]
To be clear here:

The RFC says that while servers _should_ populate post-op attrs they are not required to do so -- for example when this is very costly. The client should be able to cope without them (usually at the cost of an extra GETATTR call). So this is probably a bug in the win7 client too.

**Comment by dang** [comment]
There is a bug in fsal_ceph, but my point was that of the 3 FSALs I looked at, all 3 did it differently, so we should audit this case for all FSALs.

**Comment by uyghfjhh** [comment]
@jtlayton I merged the patch d83f9da in ganesha 2.6.3 , I can create and copy folders without errors.

**Comment by jtlayton** [comment]
Great. FWIW, it's essentially the same as the one you proposed. The only difference is that we ignore errors from the getattr since the create/setattr2 succeeded. We can still send zeroed out attrs in the case where that occurs, but it should be something quite rare.
